drivers: can: calculate a default SJW value if none is specified

henrikbrixandersen opened the pull request:
Change the CAN timing API to automatically calculate a default (Re-)Synchronization Jump Width (SJW) value if none is set by the caller. This allows automatically scaling the SJW according to the number of Time Quanta (TQ) used for phase segment 2 instead of relying on a compile-time fallback value defined in devicetree.

Update the CAN controller drivers to solely use the `sjw` and `sjw-data` devicetree properties for setting the initial timing when devicetree timing parameters are specified in Time Quanta (TQ). Any timing set via the CAN timing APIs will contain either user-provided or automatically calculated SJW values. This includes any timing parameters calculated from `bus-speed` and `bus-speed-data` devicetree properties.
    
Update the CAN controller driver tests accordingly and remove the `CAN_SJW_NO_CHANGE` definition as it has lost its meaning.

Update the descriptions for the various CAN devicetree timing properties specified in Time Quanta (TQ) to make it clear that these, if present, are only used for the initial timing parameters. Deprecate the (Re-)Synchronization Jump Width (SJW) devicetree properties for both arbitration and data phase timing as these are now only used in combination with the other TQ-based CAN timing properties, which are all deprecated.

Fixes: #63033

henrikbrixandersen committed:
drivers: can: calculate a default SJW value

Change the CAN timing calculation APIs to automatically calculate a default
(Re-)Synchronization Jump Width (SJW) value. The calculated value can be
overwritten by the caller if desired.

This allows automatically scaling the SJW according to the number of Time
Quanta (TQ) used for phase segment 2 instead of relying on a compile-time
fallback value defined in devicetree.

This reduces the can_set_timing()/can_set_timing_data() API functions to
simple setters (with validation).

Fixes: #63033

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: can: solely use sjw from devicetree for initial timing

Update the CAN controller drivers to solely use the sjw and sjw-data
devicetree properties for setting the initial timing when devicetree timing
parameters are specified in Time Quanta (TQ).

Any timing set via the CAN timing APIs will contain either user-provided or
automatically calculated SJW values. This includes any timing parameters
calculated from bus-speed and bus-speed-data devicetree properties.

Update the CAN controller driver tests accordingly and remove the
CAN_SJW_NO_CHANGE definition as it has lost its meaning.

Fixes: #63033

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
dts: bindings: can: deprecate the sjw and sjw-data properties

Update the descriptions for the various CAN devicetree timing properties
specified in Time Quanta (TQ) to make it clear that these, if present, are
only used for the initial timing parameters.

Deprecate the (Re-)Synchronization Jump Width (SJW) devicetree properties
for both arbitration and data phase timing as these are now only used in
combination with the other TQ-based CAN timing properties, which are all
deprecated.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
As an example of how this automatic scaling of the SJW impacts timing calculations, here is the output from `tests/drivers/can/timing` running on the `lpcxpresso55s36`. Notice the different value used for `sjw` at the different bitrates:

```
*** Booting Zephyr OS build zephyr-v3.4.0-4228-ge79aad5cb167 ***
Running TESTSUITE can_timing
===================================================================
testing on device can@4009d000 @ 144000000 Hz
START - test_set_timing_data_max
 PASS - test_set_timing_data_max in 0.001 seconds
===================================================================
START - test_set_timing_data_min
 PASS - test_set_timing_data_min in 0.001 seconds
===================================================================
START - test_set_timing_max
 PASS - test_set_timing_max in 0.001 seconds
===================================================================
START - test_set_timing_min
 PASS - test_set_timing_min in 0.001 seconds
===================================================================
START - test_timing
testing bitrate 20000, sample point 87.5% (valid): sjw = 18, prop_seg = 0, phase_seg1 = 251, phase_seg2 = 36, prescaler = 25 OK, sample point error 0.0%
testing bitrate 50000, sample point 87.5% (valid): sjw = 18, prop_seg = 0, phase_seg1 = 251, phase_seg2 = 36, prescaler = 10 OK, sample point error 0.0%
testing bitrate 125000, sample point 87.5% (valid): sjw = 18, prop_seg = 0, phase_seg1 = 251, phase_seg2 = 36, prescaler = 4 OK, sample point error 0.0%
testing bitrate 250000, sample point 87.5% (valid): sjw = 18, prop_seg = 0, phase_seg1 = 251, phase_seg2 = 36, prescaler = 2 OK, sample point error 0.0%
testing bitrate 500000, sample point 87.5% (valid): sjw = 18, prop_seg = 0, phase_seg1 = 251, phase_seg2 = 36, prescaler = 1 OK, sample point error 0.0%
testing bitrate 800000, sample point 80.0% (valid): sjw = 18, prop_seg = 0, phase_seg1 = 143, phase_seg2 = 36, prescaler = 1 OK, sample point error 0.0%
testing bitrate 1000000, sample point 75.0% (valid): sjw = 18, prop_seg = 0, phase_seg1 = 107, phase_seg2 = 36, prescaler = 1 OK, sample point error 0.0%
testing bitrate 125000, sample point 90.0% (valid): sjw = 6, prop_seg = 0, phase_seg1 = 114, phase_seg2 = 13, prescaler = 9 OK, sample point error 0.2%
testing bitrate 125000, sample point 80.0% (valid): sjw = 29, prop_seg = 0, phase_seg1 = 229, phase_seg2 = 58, prescaler = 4 OK, sample point error 0.2%
testing bitrate 125000, sample point 100.0% (invalid): OK
testing bitrate 1000001, sample point 75.0% (invalid): OK
 PASS - test_timing in 0.195 seconds
===================================================================
START - test_timing_data
testing bitrate 500000, sample point 87.5% (valid): sjw = 2, prop_seg = 0, phase_seg1 = 27, phase_seg2 = 4, prescaler = 9 OK, sample point error 0.0%
testing bitrate 1000000, sample point 75.0% (valid): sjw = 4, prop_seg = 0, phase_seg1 = 26, phase_seg2 = 9, prescaler = 4 OK, sample point error 0.0%
testing bitrate 500000, sample point 90.0% (valid): sjw = 2, prop_seg = 0, phase_seg1 = 31, phase_seg2 = 4, prescaler = 8 OK, sample point error 1.2%
testing bitrate 500000, sample point 80.0% (valid): sjw = 2, prop_seg = 0, phase_seg1 = 18, phase_seg2 = 5, prescaler = 12 OK, sample point error 0.9%
testing bitrate 500000, sample point 100.0% (invalid): OK
testing bitrate 8000001, sample point 75.0% (invalid): OK
 PASS - test_timing_data in 0.094 seconds
===================================================================
TESTSUITE can_timing succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [can_timing]: pass = 6, fail = 0, skip = 0, total = 6 duration = 0.293 seconds
 - PASS - [can_timing.test_set_timing_data_max] duration = 0.001 seconds
 - PASS - [can_timing.test_set_timing_data_min] duration = 0.001 seconds
 - PASS - [can_timing.test_set_timing_max] duration = 0.001 seconds
 - PASS - [can_timing.test_set_timing_min] duration = 0.001 seconds
 - PASS - [can_timing.test_timing] duration = 0.195 seconds
 - PASS - [can_timing.test_timing_data] duration = 0.094 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

alexanderwachter commented on drivers/can/can_common.c:
```diff
@@ -160,6 +160,20 @@ static int update_sampling_pnt(uint32_t total_tq, uint32_t sample_pnt,
+static void can_calc_default_sjw(struct can_timing *res, const struct can_timing *min,
+				 const struct can_timing *max)
+{
+	res->sjw = MAX(1U, MIN(res->phase_seg1, res->phase_seg2 / 2));
```
Do you need the MAX(1U, ...) if you CLAMP it later?
Do you have any reference for this calculation? Is it a recommendation from CiA?

henrikbrixandersen replied:
> Do you need the MAX(1U, ...) if you CLAMP it later?

Nice catch, no I don't. I added the `CLAMP()` call later, but didn't think of removing the `MAX()`. Removed now.

> Do you have any reference for this calculation? Is it a recommendation from CiA?

As far as I know, sadly the CiA does not give any recommendations regarding the SJW values, no. The only references, I have, are the ones mentioned in the issue (see https://github.com/zephyrproject-rtos/zephyr/issues/63033). The strongest argument to defaulting to `PHASE_SEG2 / 2` would be that the Linux kernel does this as well. You might want to read through their reasoning as well (again, linked from the issue).

henrikbrixandersen commented:
Rebased to remove unneeded call to `MAX()` spotted by @alexanderwachter.

carlescufi reviewed (dismissed):
Please add an entry in the migration guide.

cristianfunes79 commented on drivers/can/can_common.c:
```diff
@@ -372,23 +391,28 @@ int z_impl_can_set_timing_data(const struct device *dev,
 	const struct can_driver_api *api = (const struct can_driver_api *)dev->api;
 	const struct can_timing *min = can_get_timing_data_min(dev);
 	const struct can_timing *max = can_get_timing_data_max(dev);
+	struct can_timing copy = *timing_data;
```
IMO timing_data_copy seems more meaningful name than just copy.

cristianfunes79 reviewed (approved):
Added minor comment.

martinjaeger commented on drivers/can/can_mcp2515.c:
```diff
@@ -361,11 +361,8 @@ static int mcp2515_set_timing(const struct device *dev,
-		dev_data->sjw = (timing->sjw - 1) << 6;
-	}
-
-	uint8_t cnf1 = dev_data->sjw | brp;
```
As far as I can see, the `sjw` parameter is now unused and can be removed from `struct mcp2515_data` in `can_mcp2515.h`.

henrikbrixandersen replied:
Thanks, fixed.

alexanderwachter commented:
@henrikbrixandersen what do you think about having the sjw as a per mille parameter, same as the sampling point?
Phase seg2 / 2 seems somehow odd to me. This is just directly related to the sampling point. It does not make sense for me to be honest.

henrikbrixandersen committed:
doc: releases: update v3.5 migration guide with changes in CAN timing

Update the migration guide for v3.5 to mention the changes related to
setting CAN timing.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen commented:
> @henrikbrixandersen what do you think about having the sjw as a per mille parameter, same as the sampling point?

I am trying to reduce complexity, not increase it by adding yet another parameter.

> Phase seg2 / 2 seems somehow odd to me. This is just directly related to the sampling point. It does not make sense for me to be honest.

The SJW needs to be a percentage of the nominal bit time. It has to be less than or equal to both PHASE_SEG1 and PHASE_SEG2 and must be within the range of the supported SJW values for the given controller. The default value `PHASE_SEG2 / 2` may be arbitrary, but it has precedence in other implementations (e.g. the Linux kernel), which is sufficient reason for going with this as default. The user/application is still free to overwrite any default calculated SJW with whatever calculation or default value they may choose.

henrikbrixandersen commented on include/zephyr/drivers/can.h:
```diff
@@ -234,11 +234,6 @@ struct can_bus_err_cnt {
-/** Synchronization Jump Width (SJW) value to indicate that the SJW should not
- * be changed by the timing calculation.
- */
-#define CAN_SJW_NO_CHANGE 0
```
Thanks for the suggestion. Instead, as discussed elsewhere, I have gone with a simpler approach (described below).

henrikbrixandersen commented:
Thank you for the review comments, all. Based on the comment, I have tried to simplify the API even further.

I have updated this patch set to reduce the `can_set_timing()` and `can_set_timing_data()` API functions to mere settings with built-in validation of the supplied timing parameters. They no longer attempt to calculate a default SJW in case none was specified, nor do the attempt to restore a default SJW value set in devicetree. They act as setters, no more, no less.

The calculation of the default, suitable SJW has been kept in the `can_calc_timing()` and `can_calc_timing_data()` API functions. These are also used by `can_set_bitrate()`/`can_set_bitrate_data()` and thus these functions will also calculate a default SJW. The calculation is still based on `PHASE_SEG2 / 2` as this has precedence in other implementations (e.g. the Linux kernel).

henrikbrixandersen commented:
Rebased to fix merge conflict in `doc/releases/migration-guide-3.5.rst`.

henrikbrixandersen commented:
> Please add an entry in the migration guide.

Added. Please review.

alexanderwachter reviewed (changes requested):
Please give me some time today.
There are many misunderstandings what the SJW actually dose. A quick online search just proved that most people think that it is the maximum tq that the controller is allowed to shift its timing according to dominant edges on the line. This means it is a property of oscillator drifts in the network.
But I think it is the threshold  in units of tq where the receivers adds or subtracts this same value from the actual timing.
I'll continue investigating and consult the standard one again.

henrikbrixandersen commented:
> Please give me some time today.
> There are many misunderstandings what the SJW actually dose. A quick online search just proved that most people think that it is the maximum tq that the controller is allowed to shift its timing according to dominant edges on the line. This means it is a property of oscillator drifts in the network.
> But I think it is the threshold in units of tq where the receivers adds or subtracts this same value from the actual timing.
> I'll continue investigating and consult the standard one again.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/1076226/b303faaa-257f-42c1-b1ad-a872007fd3b2)

alexanderwachter commented:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/9592162/0ba9f18a-f3d1-4068-881b-f043f0ad0cd3)

That is the formula from the SJA 1000.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/9592162/9da9ef2e-c5fc-4d19-91d1-7f47d17411cd)

According to this AN it has to be as small as possible, but it must be able to compensate for the phase errors.

henrikbrixandersen commented:
> ![image](https://user-images.githubusercontent.com/9592162/271199655-0ba9f18a-f3d1-4068-881b-f043f0ad0cd3.png)
> 
> That is the formula from the SJA 1000.
> 
> ![image](https://user-images.githubusercontent.com/9592162/271199888-9da9ef2e-c5fc-4d19-91d1-7f47d17411cd.png)
> 
> According to this AN it has to be as small as possible, but it must be able to compensate for the phase errors.

For classic CAN 2.0b CAN controllers like the aging SJA1000, the implemented algorithm for calculating SJW still honours this. Example `tests/drivers/can/timing/` output on `qemu_x86` using the SJA1000 backend:

```
SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
Booting from ROM..
*** Booting Zephyr OS build zephyr-v3.4.0-4479-g8d88f9f891d3 ***
Running TESTSUITE can_timing
===================================================================
testing on device can0 @ 8000000 Hz
START - test_set_timing_data_max
 SKIP - test_set_timing_data_max in 0.001 seconds
===================================================================
START - test_set_timing_data_min
 SKIP - test_set_timing_data_min in 0.001 seconds
===================================================================
START - test_set_timing_max
 PASS - test_set_timing_max in 0.001 seconds
===================================================================
START - test_set_timing_min
 PASS - test_set_timing_min in 0.001 seconds
===================================================================
START - test_timing
testing bitrate 20000, sample point 87.5% (valid): sjw = 1, prop_seg = 0, phase_seg1 = 13, phase_seg2 = 2, prescaler = 25 OK, sample point error 0.0%
testing bitrate 50000, sample point 87.5% (valid): sjw = 1, prop_seg = 0, phase_seg1 = 13, phase_seg2 = 2, prescaler = 10 OK, sample point error 0.0%
testing bitrate 125000, sample point 87.5% (valid): sjw = 1, prop_seg = 0, phase_seg1 = 13, phase_seg2 = 2, prescaler = 4 OK, sample point error 0.0%
testing bitrate 250000, sample point 87.5% (valid): sjw = 1, prop_seg = 0, phase_seg1 = 13, phase_seg2 = 2, prescaler = 2 OK, sample point error 0.0%
testing bitrate 500000, sample point 87.5% (valid): sjw = 1, prop_seg = 0, phase_seg1 = 13, phase_seg2 = 2, prescaler = 1 OK, sample point error 0.0%
testing bitrate 800000, sample point 80.0% (valid): sjw = 1, prop_seg = 0, phase_seg1 = 7, phase_seg2 = 2, prescaler = 1 OK, sample point error 0.0%
testing bitrate 1000000, sample point 75.0% (valid): sjw = 1, prop_seg = 0, phase_seg1 = 5, phase_seg2 = 2, prescaler = 1 OK, sample point error 0.0%
testing bitrate 125000, sample point 90.0% (valid): sjw = 1, prop_seg = 0, phase_seg1 = 13, phase_seg2 = 2, prescaler = 4 OK, sample point error 2.5%
testing bitrate 125000, sample point 80.0% (valid): sjw = 2, prop_seg = 0, phase_seg1 = 11, phase_seg2 = 4, prescaler = 4 OK, sample point error 5.0%
testing bitrate 125000, sample point 100.0% (invalid): OK
testing bitrate 1000001, sample point 75.0% (invalid): OK
 PASS - test_timing in 0.011 seconds
===================================================================
START - test_timing_data
 SKIP - test_timing_data in 0.001 seconds
===================================================================
TESTSUITE can_timing succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [can_timing]: pass = 3, fail = 0, skip = 3, total = 6 duration = 0.016 seconds
 - SKIP - [can_timing.test_set_timing_data_max] duration = 0.001 seconds
 - SKIP - [can_timing.test_set_timing_data_min] duration = 0.001 seconds
 - PASS - [can_timing.test_set_timing_max] duration = 0.001 seconds
 - PASS - [can_timing.test_set_timing_min] duration = 0.001 seconds
 - PASS - [can_timing.test_timing] duration = 0.011 seconds
 - SKIP - [can_timing.test_timing_data] duration = 0.001 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

alexanderwachter commented:
Also calculators found online

https://www.kvaser.com/support/calculators/can-fd-bit-timing-calculator/

https://www.ti.com/lit/an/sprac35/sprac35.pdf?ts=1695884979052&ref_url=https%253A%252F%252Fwww.google.com%252F

Only depend on the oscillator tolerance.
None of the documentation/calculators depend on the sampling point.
So in my opinion, we need a parameter for the allowed oscillator tolerances in the network.

henrikbrixandersen commented:
> So in my opinion, we need a parameter for the allowed oscillator tolerances in the network.

Thank you for sharing your opinion on this and sharing the various sources. There is no doubt that the SJW is there to allow compensating for oscillator/CAN clock tolerances between the different nodes in the network (that is what this entire PR/RFC is about). This PR deals with calculating a _default_, _suitable_ SJW to solve most use-cases and to avoid communication errors (due to phase errors) out-of-the-box when people use CAN on Zephyr.

This is only a default SJW (similar as to how we have default sample points, but applications can override them). Users can _still_ calculate their own SJW based on their knowledge about oscillator tolerance across all nodes in _their_ particular CAN network, if these are known beforehand. For many use-cases, the oscillator tolerances between all nodes is, however, not known by the firmware vendor. This goes for simple test setups as well.

I am convinced that we do not need yet another parameter for setting allowed oscillator tolerances in the Zephyr CAN API. We already support this by allowing users/applications to set a custom SJW, which is how the CAN standard expresses this tolerance. Our _current_ solution, however, fails to scale this and results in out-of-the-box communication errors (especially when using CAN-FD) when communicating with off-the-shelf CAN equipment (which, by the way, often sets the SJW to a much less conservative value than `PHASE_SEG2 / 2`).

Did you read through the argumentation put forward when this exact same change was introduced in the Linux kernel (see https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=80bcf5ec9927f0a8056495d746b74f57b1e1ad8b)?

alexanderwachter commented:
I can understand the argument of a default value and agree.
But I think that basing the default on the sampling point, which seems to be unrelated, does not make much sense.
What do you think of providing a good gefault tolerance? And maybe a helper to get a proper SJW if the tolerances are known by the user?

henrikbrixandersen commented:
> I can understand the argument of a default value and agree.

Good.

> But I think that basing the default on the sampling point, which seems to be unrelated, does not make much sense.
> What do you think of providing a good gefault tolerance? And maybe a helper to get a proper SJW if the tolerances are known by the user?

You shouldn't see it as much as being calculated based on the sample point, more as being calculated relative to the nominal bit time expressed in TQ (as `PHASE_SEG2` is relative to this and the used sample point). Since the SJW can _never_ exceed the value of neither `PHASE_SEG1` nor `PHASE_SEG2` (as these are the segments that can be shortened/lengthened by the SJW value), it makes sense to only look at these (the calculation defaults to `PHASE_SEG2 / 2`, but caps it to `PHASE_SEG1` as well. The sample point in any CAN network will likely not deviated much from the 75% to 90% range, so the impact on the calculated, default SJW by moving the sample point slightly is kept at a minimum.

alexanderwachter commented:
> > I can understand the argument of a default value and agree.
> 
> 
> 
> Good.
> 
> 
> 
> > But I think that basing the default on the sampling point, which seems to be unrelated, does not make much sense.
> 
> > What do you think of providing a good gefault tolerance? And maybe a helper to get a proper SJW if the tolerances are known by the user?
> 
> 
> 
> You shouldn't see it as much as being calculated based on the sample point, more as being calculated relative to the nominal bit time expressed in TQ (as `PHASE_SEG2` is relative to this and the used sample point). Since the SJW can _never_ exceed the value of neither `PHASE_SEG1` nor `PHASE_SEG2` (as these are the segments that can be shortened/lengthened by the SJW value), it makes sense to only look at these (the calculation defaults to `PHASE_SEG2 / 2`, but caps it to `PHASE_SEG1` as well). The sample point in any CAN network will likely not deviated much from the 75% to 90% range, so the impact on the calculated, default SJW by moving the sample point slightly is kept at a minimum.

I understand that. But still it can change the SWJ by just changing an unrelated parameter.
Using a good default of lets say 1% (500ppm on both nodes) would just be more correct.

alexanderwachter commented:
Another point is that some controllers include the prop_seg in the seg_1 and some have it separate. So the sjw could also be influenced by the Controller used. Taking a tq and multiply it by a factor resulting from the calculations in the documents shared would not suffer from that.

alexanderwachter commented:
I think we need to change

https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/can/can_common.c line 141.

To not limit the SJW it makes sense to put more tq into segment 1 rather than prop seg.
What do you think @henrikbrixandersen

henrikbrixandersen commented:
> I think we need to change
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/can/can_common.c line 141.
> 
> To not limit the SJW it makes sense to put more tq into segment 1 rather than prop seg. What do you think @henrikbrixandersen

Please open a separate issue for that. This is not related to calculating a default SJW. The limitation on the allowed range for SJW is present regardless.

henrikbrixandersen commented:
> Using a good default of lets say 1% (500ppm on both nodes) would just be more correct.

1% of what? The local CAN clock? I cannot see how you can calculate it like that?

Say you have two CAN nodes on the same CAN bus. One node is using a CAN clock of 80MHz, the other a CAN clock of 16MHz. 500ppm tolerance for the first would be +/-40kHz, for the latter it would be +/-8kHz. How would the _default_ calculation be able to take the tolerance of _all_ the nodes on the CAN bus into account, giving their potentially different CAN clock rates (and indeed tolerances)?

alexanderwachter commented:
> > Using a good default of lets say 1% (500ppm on both nodes) would just be more correct.
> 
> 
> 
> 1% of what? The local CAN clock? I cannot see how you can calculate it like that?
> 
> 
> 
> Say you have two CAN nodes on the same CAN bus. One node is using a CAN clock of 80MHz, the other a CAN clock of 16MHz. 500ppm tolerance for the first would be +/-40kHz, for the latter it would be +/-8kHz. How would the _default_ calculation be able to take the tolerance of _all_ the nodes on the CAN bus into account, giving their potentially different CAN clock rates (and indeed tolerances)?

Since the suggestion is in percent (or ppm), it is relative and does not depend on a specific value. It also does not matter which node runs on what frequency or has more drift. It is just important hat the absolute value of both drifts is not larger than 1% of the nominal time.


![image](https://github.com/zephyrproject-rtos/zephyr/assets/9592162/a2e4e798-5879-4d4d-8531-80faf5128190)

https://ww1.microchip.com/downloads/en/Appnotes/00754.pdf

The only thing we need to know is how many tq we have in a bit (1 + prop + seg1 + seg2) and divide it with the factor we need to agree on. And then cap it with 1, seg1, seg2.

henrikbrixandersen commented:
> The only thing we need to know is how many tq we have in a bit (1 + prop + seg1 + seg2) and divide it with the factor we need to agree on. And then cap it with 1, seg1, seg2.

You do notice how all the calculators, chip-specific applications notes, etc. you've linked to far takes the CAN clock + tolerance of _all_ nodes on the CAN bus into account when calculating this, right?

alexanderwachter commented:
Yes, I do. We agreed on not introducing a new parameter.
Thats why we need a good default value and assume that all nodes are within some range. 1% was just a proposal.

henrikbrixandersen commented:
I feel like we're talking past each other somehow. 

Take for instance the CAN-FD timing example from https://www.can-cia.org/fileadmin/resources/documents/proceedings/2013_hartwich_v2.pdf :

![image](https://github.com/zephyrproject-rtos/zephyr/assets/1076226/0e4b5a65-490a-4fce-91f7-e15228540b74)

Notice how the proposed SJW for the arbitration and data phases does not change with the changes in bit time, it does not change with the changes to the amount of TQ in the bit time, nor is it the same across the two examples although they use the same CAN clock and the same clock tolerance.

I still believe the default of `PHASE_SEG2 / 2` is an adequate, conservative default.

alexanderwachter commented:
> > I still believe that the sp should have nothing to do with the sjw.
> 
> Please see https://github.com/zephyrproject-rtos/zephyr/pull/63034#issuecomment-1738667806 where I already answered this. 

I read and understood. Still the sp is influencing the value.

What I could agree on is that the sp is somehow a hint that here is long propagation delay expected if the sp is chosen very late. But for the dataphase the propagation delay is not relevant for sp. This means that the seg1 value could be very large and does not cap the SWJ.

alexanderwachter reviewed (approved):
I'll be not available for the next few days. Approving to unblock (i have no idea how to remove a review in the mobile app -> approve)
I think this version of calculating the SJW is definitely better than just using 1 and is definitely not totally wrong. But i think it is not perfect.

martinjaeger reviewed (approved):
Thanks for the updates and further explanations. I like the final solution, which eliminates the magic `0` for no change or automatic calculation. The `set` and `calc` functions are now more clear.

If necessary, the calculation of the default SJW can be further improved in a future PR, even though I think taking the same approach as Linux does makes sense for now.
